Preserve symlinks in streaming interface

shanemcd opened the pull request.

shanemcd committed:
Preserve symlinks in streaming interface

chrismeyersfsu commented on ansible_runner/utils/streaming.py:
```diff
@@ -19,14 +20,21 @@ def stream_dir(source_directory, stream):
+                            file_relative_path = os.path.join(archive_relative_path, fname)
+                            zip_info = zipfile.ZipInfo(file_relative_path)
+                            zip_info.create_system = 3
+                            permissions = 0o777
```
Do unzipped files result in permission `777` ?

shanemcd replied:
See my comment below.

chrismeyersfsu commented on ansible_runner/utils/streaming.py:
```diff
@@ -64,5 +72,11 @@ def unstream_dir(stream, length, target_directory):
+                if mode[:1] == 'l':
+                    link = open(out_path).read()
+                    os.remove(out_path)
+                    os.symlink(link, out_path)
```
We want to retain the perms from the original project, even on the symlink.

shanemcd replied:
I think this is just how symlinks work:

```
$ cd /tmp/
$ ln -sf /etc/passwd /tmp/hello
$ sudo su - anotheruser
Last login: Thu Jun  3 15:34:33 EDT 2021 on pts/27
$ ls -la /tmp/hello
lrwxrwxrwx. 1 shanemcd shanemcd 11 Jun  3 15:35 /tmp/hello -> /etc/passwd
$ rm /tmp/hello
rm: cannot remove '/tmp/hello': Operation not permitted
```

AlanCoding committed:
Add test coverage for transmission symlink preservation

this tests the utils methods that are used for
  ansible-runner transmit
  ansible-runner worker

which are used to make a zipfile of the private_data_dir
and then stream it over the wire

chrismeyersfsu commented on ansible_runner/utils/streaming.py:
```diff
@@ -64,5 +72,11 @@ def unstream_dir(stream, length, target_directory):
+                perms = info.external_attr >> 16
+                mode = stat.filemode(perms)
+                if mode[:1] == 'l':
+                    link = open(out_path).read()
```
could OOM if a malicious user puts a bunch of content in the symlink data portion. There are probably other/better way to OOM the system than that. Besides, the project SCM should be scrutinized via review process "why has our git checkout time jumped from 30 seconds to 3 hours?"

shanemcd replied:
https://unix.stackexchange.com/questions/32795/what-is-the-maximum-allowed-filename-and-folder-size-with-ecryptfs

>Linux has a maximum filename length of 255 characters for most filesystems (including EXT4), and a maximum path of 4096 characters.

shanemcd committed:
Improve logic around extracting symlinks from zip archives

Co-authored-by: Jeff Bradberry <jeff.bradberry@gmail.com>

AlanCoding committed:
Add transmit stream test for file permissions and rewriting

AlanCoding reviewed (approved):
I contributed the tests, but did not modify the main source tree here.

All the specific concerns that were raised have been given test coverage. @chrismeyersfsu raised some points about file permissions, which have been addressed and tested to my satisfaction. Other points made here seem to be addressed as well.

New tests are passing locally, I'm happy

chrismeyersfsu commented on ansible_runner/utils/streaming.py:
```diff
@@ -62,7 +70,20 @@ def unstream_dir(stream, length, target_directory):
+                
+                is_symlink = mode[:1] == 'l'
+                if is_symlink and os.path.exists(out_path):
+                    os.remove(out_path)
```
I wouldn't think we need this remove since we remove below and I wouldn't expect `out_path` to ever exist in this case.
